File: packages/manifest-schema/src/validator.ts
/**
 * Pure validation engine for Software Capsule Manifests.
 * Implements syntax, schema, and semantic checks per docs/manifest-spec/ and docs/api-cli-spec/.
 */
import Ajv2020 from "ajv/dist/2020.js";
import addFormats from "ajv-formats";
import schemaJson from "../schema/capsule.manifest.schema.json" with { type: "json" };
import type {
  CapsuleManifest,
  ValidationCheck,
  ValidationResult,
  ConnectorDeclaration,
} from "./types.js";

// Handle CJS/ESM interop under NodeNext module resolution
const Ajv2020Constructor: any = (Ajv2020 as any).default || Ajv2020;
const addFormatsFn: any = (addFormats as any).default || addFormats;

const ajv = new Ajv2020Constructor({ allErrors: true, strict: false });
addFormatsFn(ajv);
const validateSchema = ajv.compile(schemaJson);

const HOSTNAME_REGEX =
  /^(?=.{1,253}$)(?!-)[A-Za-z0-9-]{1,63}(?<!-)(\.(?!-)[A-Za-z0-9-]{1,63}(?<!-))*$/;
const IP_REGEX =
  /^(?:\d{1,3}\.){3}\d{1,3}$|^([0-9a-fA-F]{1,4}:){7}[0-9a-fA-F]{1,4}$/;
const ID_REGEX = /^[a-z0-9][a-z0-9-]{0,62}$/;
const ROLE_REGEX = /^[a-z][a-z0-9_-]{0,31}$/;

export function validateManifest(input: unknown): ValidationResult {
  const checks: ValidationCheck[] = [];
  const requiredApprovals: string[] = [];

  // 1. Basic Type Check
  if (!input || typeof input !== "object" || Array.isArray(input)) {
    const check: ValidationCheck = {
      name: "input_format",
      status: "fail",
      code: "invalid_manifest",
      path: null,
      message: "Manifest must be a non-null object.",
      hint: "Provide a valid YAML/JSON document with top-level fields: apiVersion, id, name, shape, runtime.",
    };
    return {
      valid: false,
      errors: [check],
      checks: [check],
      required_approvals: [],
      warnings: [],
    };
  }

  const manifest = input as Record<string, any>;

  // 2. Schema Validation (Draft 2020-12)
  const isSchemaValid = validateSchema(manifest);
  if (!isSchemaValid && validateSchema.errors) {
    for (const err of validateSchema.errors) {
      let path = err.instancePath
        ? err.instancePath.replace(/^\//, "").replace(/\//g, ".")
        : null;
      let message = err.message || "Schema validation error";
      let hint = "Fix the field according to capsule.manifest.schema.json.";

      if (err.keyword === "required") {
        const missing = (err.params as any).missingProperty;
        path = path ? `${path}.${missing}` : missing;
        message = `Missing required property '${missing}'.`;
        hint = `Add the required '${missing}' property to your manifest.`;
      } else if (err.keyword === "additionalProperties") {
        const extra = (err.params as any).additionalProperty;
        path = path ? `${path}.${extra}` : extra;
        message = `Unrecognized property '${extra}'.`;
        hint = `Remove '${extra}'. Unknown properties are not permitted.`;
      } else if (err.keyword === "const") {
        const expected = (err.params as any).allowedValue;
        message = `Value must be '${expected}'.`;
        hint = `Change value to '${expected}'.`;
      } else if (err.keyword === "pattern") {
        if (path === "id") {
          message = `Invalid capsule ID '${manifest.id}'.`;
          hint =
            "Capsule ID must be 1-63 lowercase alphanumeric characters and hyphens (e.g. leave-tracker).";
        } else if (path?.startsWith("roles")) {
          message = `Invalid role name format.`;
          hint =
            "Role names must start with lowercase letter and use snake_case (e.g. hr_manager).";
        } else {
          message = `Value at '${path}' does not match required pattern.`;
          hint = `Ensure '${path}' conforms to pattern: ${(err.params as any).pattern}`;
        }
      }

      checks.push({
        name: "schema_validation",
        status: "fail",
        code: "invalid_manifest",
        path,
        message,
        hint,
      });
    }
  } else {
    checks.push({
      name: "schema_validation",
      status: "pass",
      code: "schema_valid",
      message: "Manifest conforms to JSON Schema Draft 2020-12.",
      path: null,
    });
  }

  // 3. Semantic Checks: Shape
  if (manifest.shape !== undefined) {
    if (manifest.shape !== "web-app") {
      checks.push({
        name: "shape_check",
        status: "fail",
        code: "unsupported_shape",
        path: "shape",
        message: `Unsupported shape '${manifest.shape}'.`,
        hint: "Set 'shape' to 'web-app'. Only web-app is supported in Alpha/MVP.",
      });
    } else {
      checks.push({
        name: "shape_check",
        status: "pass",
        code: "shape_valid",
        message: "Shape 'web-app' is valid.",
        path: "shape",
      });
    }
  }

  // 4. Semantic Checks: Runtime
  if (manifest.runtime !== undefined) {
    if (manifest.runtime !== "node22") {
      checks.push({
        name: "runtime_check",
        status: "fail",
        code: "unsupported_runtime",
        path: "runtime",
        message: `Unsupported runtime '${manifest.runtime}'.`,
        hint: "Set 'runtime' to 'node22'. Only node22 is supported in Alpha/MVP.",
      });
    } else {
      checks.push({
        name: "runtime_check",
        status: "pass",
        code: "runtime_valid",
        message: "Runtime 'node22' is valid.",
        path: "runtime",
      });
    }
  }

  // 5. Semantic Checks: ID Format
  if (manifest.id !== undefined && typeof manifest.id === "string") {
    if (!ID_REGEX.test(manifest.id)) {
      checks.push({
        name: "id_check",
        status: "fail",
        code: "invalid_manifest",
        path: "id",
        message: `Invalid capsule ID '${manifest.id}'.`,
        hint: "Capsule ID must be 1-63 lowercase alphanumeric characters and hyphens (e.g. leave-tracker).",
      });
    } else {
      checks.push({
        name: "id_check",
        status: "pass",
        code: "id_valid",
        message: "Capsule ID format is valid.",
        path: "id",
      });
    }
  }

  // 6. Semantic Checks: Roles Uniqueness & Format
  if (Array.isArray(manifest.roles)) {
    const seenRoles = new Set<string>();
    for (let i = 0; i < manifest.roles.length; i++) {
      const role = manifest.roles[i];
      if (typeof role === "string") {
        if (!ROLE_REGEX.test(role)) {
          checks.push({
            name: "roles_check",
            status: "fail",
            code: "invalid_manifest",
            path: `roles[${i}]`,
            message: `Invalid role format '${role}'.`,
            hint: "Roles must start with a letter and contain only lowercase letters, digits, underscores, or hyphens (max 32 chars).",
          });
        }
        if (seenRoles.has(role)) {
          checks.push({
            name: "roles_check",
            status: "fail",
            code: "invalid_manifest",
            path: `roles[${i}]`,
            message: `Duplicate role '${role}'.`,
            hint: "Remove duplicate role declarations.",
          });
        }
        seenRoles.add(role);
      }
    }
  }

  // 7. Semantic Checks: Network Egress & Default-Deny
  const egressList = manifest.egress !== undefined ? manifest.egress : [];
  if (Array.isArray(egressList)) {
    if (egressList.length === 0) {
      checks.push({
        name: "egress_check",
        status: "pass",
        code: "egress_default_deny",
        message: "Default-deny egress policy is active.",
        path: "egress",
      });
    } else {
      for (let i = 0; i < egressList.length; i++) {
        const dest = egressList[i];
        if (typeof dest !== "string") continue;

        // Check for URL scheme
        if (dest.includes("://") || dest.includes("/")) {
          checks.push({
            name: "egress_check",
            status: "fail",
            code: "egress_denied",
            path: `egress[${i}]`,
            message: `Egress destination '${dest}' contains a URL scheme or path.`,
            hint: "Specify only the FQDN hostname (e.g. api.example.com). Protocols, ports, and paths are not allowed.",
          });
          continue;
        }

        // Check for port
        if (dest.includes(":")) {
          checks.push({
            name: "egress_check",
            status: "fail",
            code: "egress_denied",
            path: `egress[${i}]`,
            message: `Egress destination '${dest}' contains a port number.`,
            hint: "Specify only the FQDN hostname without port.",
          });
          continue;
        }

        // Check for direct IP address
        if (IP_REGEX.test(dest)) {
          checks.push({
            name: "egress_check",
            status: "fail",
            code: "egress_denied",
            path: `egress[${i}]`,
            message: `Direct IP address destination '${dest}' is prohibited.`,
            hint: "Specify an approved external domain hostname instead of a raw IP address.",
          });
          continue;
        }

        // Check for loopback / local
        const lower = dest.toLowerCase();
        if (
          lower === "localhost" ||
          lower.endsWith(".localhost") ||
          lower === "metadata.google.internal"
        ) {
          checks.push({
            name: "egress_check",
            status: "fail",
            code: "egress_denied",
            path: `egress[${i}]`,
            message: `Loopback or metadata destination '${dest}' is prohibited.`,
            hint: "Loopback and internal cloud metadata addresses cannot be declared as egress destinations.",
          });
          continue;
        }

        // Validate hostname syntax
        if (!HOSTNAME_REGEX.test(dest)) {
          checks.push({
            name: "egress_check",
            status: "fail",
            code: "egress_denied",
            path: `egress[${i}]`,
            message: `Egress destination '${dest}' is not a valid hostname.`,
            hint: "Provide a valid fully qualified domain name (e.g. api.slack.com).",
          });
          continue;
        }

        checks.push({
          name: "egress_check",
          status: "pass",
          code: "egress_allowed",
          message: `Egress destination '${dest}' passed hostname validation.`,
          path: `egress[${i}]`,
        });
      }
    }
  }

  // 8. Semantic Checks: Resource Limits
  if (manifest.limits && typeof manifest.limits === "object") {
    const limits = manifest.limits;
    if (
      limits.memory_mb !== undefined &&
      (limits.memory_mb < 64 || limits.memory_mb > 16384)
    ) {
      checks.push({
        name: "limits_check",
        status: "fail",
        code: "quota_exceeded",
        path: "limits.memory_mb",
        message: `Requested memory_mb (${limits.memory_mb}) is outside allowed range (64 - 16384 MB).`,
        hint: "Set memory_mb between 64 and 16384 (e.g. 256).",
      });
    }
    if (
      limits.request_timeout_s !== undefined &&
      (limits.request_timeout_s < 1 || limits.request_timeout_s > 300)
    ) {
      checks.push({
        name: "limits_check",
        status: "fail",
        code: "quota_exceeded",
        path: "limits.request_timeout_s",
        message: `Request timeout (${limits.request_timeout_s}s) is outside allowed range (1 - 300 seconds).`,
        hint: "Set request_timeout_s between 1 and 300 (e.g. 30).",
      });
    }
    if (
      limits.max_active_instances !== undefined &&
      limits.max_active_instances !== 1
    ) {
      checks.push({
        name: "limits_check",
        status: "fail",
        code: "quota_exceeded",
        path: "limits.max_active_instances",
        message: `max_active_instances must be 1.`,
        hint: "SQLite write model requires max_active_instances: 1.",
      });
    }
  }

  // 9. Semantic Checks: Capabilities & Connectors Defaults / Approvals
  if (manifest.capabilities && typeof manifest.capabilities === "object") {
    const caps = manifest.capabilities;
    if (caps.db && caps.db.type !== "sqlite") {
      checks.push({
        name: "capabilities_check",
        status: "fail",
        code: "capability_not_allowed",
        path: "capabilities.db.type",
        message: `Database type '${caps.db.type}' is not supported.`,
        hint: "Set capabilities.db.type to 'sqlite'.",
      });
    }

    if (Array.isArray(caps.connectors)) {
      for (let i = 0; i < caps.connectors.length; i++) {
        const conn = caps.connectors[i];
        if (!conn || typeof conn !== "object") continue;

        const actsAs = conn.acts_as || "viewer";
        if (actsAs === "service") {
          const approvalKey = `connectors.${conn.name}:service`;
          requiredApprovals.push(approvalKey);
          checks.push({
            name: "connector_approval",
            status: "warn",
            code: "approval_required",
            path: `capabilities.connectors[${i}].acts_as`,
            message: `Privileged 'service' identity requested for connector '${conn.name}'.`,
            hint: "Service identity requires human owner approval before activation.",
          });
        }
      }
    }
  }

  // 10. Semantic Checks: Schedule (Deferred to Phase 3)
  if (
    manifest.schedule &&
    Array.isArray(manifest.schedule) &&
    manifest.schedule.length > 0
  ) {
    checks.push({
      name: "schedule_check",
      status: "warn",
      code: "unsupported_capability",
      path: "schedule",
      message:
        "Scheduled execution is deferred until Phase 3 and is not active in Alpha.",
      hint: "Remove the 'schedule' field or note that it will remain dormant until Phase 3.",
    });
  }

  // Compile overall status
  const errors = checks.filter((c) => c.status === "fail");
  const warnings = checks.filter((c) => c.status === "warn");
  const valid = errors.length === 0;

  // Compute effective manifest with defaults resolved
  let effectiveManifest: CapsuleManifest | undefined = undefined;
  if (valid) {
    const normalizedConnectors: ConnectorDeclaration[] | undefined = manifest
      .capabilities?.connectors
      ? manifest.capabilities.connectors.map((c: any) => ({
          name: c.name,
          channel: c.channel,
          acts_as: c.acts_as || "viewer",
        }))
      : undefined;

    effectiveManifest = {
      apiVersion: manifest.apiVersion,
      id: manifest.id,
      name: manifest.name,
      shape: manifest.shape,
      runtime: manifest.runtime,
      roles: manifest.roles ? [...manifest.roles] : undefined,
      capabilities: manifest.capabilities
        ? {
            ...manifest.capabilities,
            connectors: normalizedConnectors,
          }
        : undefined,
      egress: manifest.egress ? [...manifest.egress] : [],
      sharing: manifest.sharing || { default: "org" },
      limits: {
        cpu: manifest.limits?.cpu || "small",
        memory_mb: manifest.limits?.memory_mb || 256,
        request_timeout_s: manifest.limits?.request_timeout_s || 30,
        db_max_mb: manifest.limits?.db_max_mb || 500,
        blob_max_mb: manifest.limits?.blob_max_mb || 200,
        max_active_instances: 1,
      },
    };
  }

  return {
    valid,
    errors,
    checks,
    required_approvals: requiredApprovals,
    warnings,
    effective_manifest: effectiveManifest,
  };
}

/**
 * Helper to check if a shape is currently supported.
 */
export function isManifestShapeValid(shape: string): boolean {
  return shape === "web-app";
}

/**
 * Helper to check if a runtime is currently supported.
 */
export function isRuntimeValid(runtime: string): boolean {
  return runtime === "node22";
}

File: services/control-plane/src/manifest/validator.py
"""
Manifest Validator for Software Capsules in Control Plane.
Enforces JSON Schema Draft 2020-12 and semantic checks identical to @capsule/manifest-schema.
"""
import json
import re
from pathlib import Path
from typing import Any, Dict, List, Optional
import jsonschema
from jsonschema import Draft202012Validator

HOSTNAME_REGEX = re.compile(
    r"^(?=.{1,253}$)(?!-)[A-Za-z0-9-]{1,63}(?<!-)(\.(?!-)[A-Za-z0-9-]{1,63}(?<!-))*$"
)
IP_REGEX = re.compile(r"^(?:\d{1,3}\.){3}\d{1,3}$|^([0-9a-fA-F]{1,4}:){7}[0-9a-fA-F]{1,4}$")
ID_REGEX = re.compile(r"^[a-z0-9][a-z0-9-]{0,62}$")
ROLE_REGEX = re.compile(r"^[a-z][a-z0-9_-]{0,31}$")

# Find and load the shared schema
def _load_schema() -> Dict[str, Any]:
    candidates = [
        Path(__file__).resolve().parents[4] / "packages" / "manifest-schema" / "schema" / "capsule.manifest.schema.json",
        Path(__file__).resolve().parents[3] / "packages" / "manifest-schema" / "schema" / "capsule.manifest.schema.json",
        Path("packages/manifest-schema/schema/capsule.manifest.schema.json").resolve(),
    ]
    for p in candidates:
        if p.is_file():
            return json.loads(p.read_text(encoding="utf-8"))
    raise FileNotFoundError("Could not locate capsule.manifest.schema.json")

SCHEMA_JSON = _load_schema()
SCHEMA_VALIDATOR = Draft202012Validator(SCHEMA_JSON)


def validate_manifest(input_data: Any) -> Dict[str, Any]:
    """Pure validation function for Capsule manifests."""
    checks: List[Dict[str, Any]] = []
    required_approvals: List[str] = []
    warnings: List[str] = []

    # 1. Basic Type Check
    if not isinstance(input_data, dict):
        check = {
            "name": "basic_type",
            "status": "fail",
            "code": "invalid_manifest",
            "path": None,
            "message": "Manifest must be a non-null object.",
            "hint": "Provide a valid YAML/JSON document with top-level fields: apiVersion, id, name, shape, runtime.",
        }
        return {
            "valid": False,
            "errors": [check],
            "checks": [check],
            "required_approvals": [],
            "warnings": [],
            "effective_manifest": None,
        }

    manifest = dict(input_data)

    # 2. Schema Validation (Draft 2020-12)
    schema_errors = sorted(SCHEMA_VALIDATOR.iter_errors(manifest), key=lambda e: e.path)
    if schema_errors:
        for err in schema_errors:
            path = ".".join(str(p) for p in err.path) if err.path else None
            message = err.message
            hint = "Fix the field according to capsule.manifest.schema.json."

            if err.validator == "required":
                missing = err.validator_value[0] if isinstance(err.validator_value, list) else str(err.message)
                # match property name from message e.g. "'id' is a required property"
                match = re.search(r"'([^']+)' is a required property", err.message)
                if match:
                    missing = match.group(1)
                path = f"{path}.{missing}" if path else missing
                message = f"Missing required property '{missing}'."
                hint = f"Add the required '{missing}' property to your manifest."
            elif err.validator == "additionalProperties":
                match = re.search(r"Additional properties are not allowed \('([^']+)'", err.message)
                extra = match.group(1) if match else "unknown"
                path = f"{path}.{extra}" if path else extra
                message = f"Unrecognized property '{extra}'."
                hint = f"Remove '{extra}'. Unknown properties are not permitted."
            elif err.validator == "const":
                expected = err.validator_value
                message = f"Value must be '{expected}'."
                hint = f"Change value to '{expected}'."
            elif err.validator == "pattern":
                if path == "id":
                    message = f"Invalid capsule ID '{manifest.get('id')}'."
                    hint = "Capsule ID must be 1-63 lowercase alphanumeric characters and hyphens (e.g. leave-tracker)."
                elif path and path.startswith("roles"):
                    message = "Invalid role name format."
                    hint = "Role names must start with lowercase letter and use snake_case (e.g. hr_manager)."
                else:
                    message = f"Value at '{path}' does not match required pattern."
                    hint = f"Ensure '{path}' conforms to pattern: {err.validator_value}"

            checks.append({
                "name": "schema_validation",
                "status": "fail",
                "code": "invalid_manifest",
                "path": path,
                "message": message,
                "hint": hint,
            })
    else:
        checks.append({
            "name": "schema_validation",
            "status": "pass",
            "code": "schema_valid",
            "path": None,
            "message": "Manifest conforms to JSON Schema Draft 2020-12.",
        })

    # 3. Semantic Checks: Shape
    shape = manifest.get("shape")
    if shape is not None:
        if shape != "web-app":
            checks.append({
                "name": "shape_check",
                "status": "fail",
                "code": "unsupported_shape",
                "path": "shape",
                "message": f"Unsupported shape '{shape}'.",
                "hint": "Set 'shape' to 'web-app'. Only web-app is supported in Alpha/MVP.",
            })
        else:
            checks.append({
                "name": "shape_check",
                "status": "pass",
                "code": "shape_valid",
                "path": "shape",
                "message": "Shape 'web-app' is valid.",
            })

    # 4. Semantic Checks: Runtime
    runtime = manifest.get("runtime")
    if runtime is not None:
        if runtime != "node22":
            checks.append({
                "name": "runtime_check",
                "status": "fail",
                "code": "unsupported_runtime",
                "path": "runtime",
                "message": f"Unsupported runtime '{runtime}'.",
                "hint": "Set 'runtime' to 'node22'. Only node22 is supported in Alpha/MVP.",
            })
        else:
            checks.append({
                "name": "runtime_check",
                "status": "pass",
                "code": "runtime_valid",
                "path": "runtime",
                "message": "Runtime 'node22' is valid.",
            })

    # 5. Semantic Checks: ID Format
    cid = manifest.get("id")
    if cid is not None and isinstance(cid, str):
        if not ID_REGEX.match(cid):
            checks.append({
                "name": "id_check",
                "status": "fail",
                "code": "invalid_manifest",
                "path": "id",
                "message": f"Invalid capsule ID '{cid}'.",
                "hint": "Capsule ID must be 1-63 lowercase alphanumeric characters and hyphens (e.g. leave-tracker).",
            })
        else:
            checks.append({
                "name": "id_check",
                "status": "pass",
                "code": "id_valid",
                "path": "id",
                "message": "Capsule ID format is valid.",
            })

    # 6. Semantic Checks: Roles Uniqueness & Format
    roles = manifest.get("roles")
    declared_roles = set()
    if isinstance(roles, list):
        seen_roles = set()
        for i, role in enumerate(roles):
            if isinstance(role, str):
                if not ROLE_REGEX.match(role):
                    checks.append({
                        "name": f"role_format_{i}",
                        "status": "fail",
                        "code": "invalid_manifest",
                        "path": f"roles.{i}",
                        "message": f"Invalid role name '{role}'.",
                        "hint": "Role names must start with lowercase letter and use snake_case (e.g. hr_manager).",
                    })
                if role in seen_roles:
                    checks.append({
                        "name": f"role_unique_{i}",
                        "status": "fail",
                        "code": "invalid_manifest",
                        "path": f"roles.{i}",
                        "message": f"Duplicate role '{role}'.",
                        "hint": f"Remove duplicate role '{role}' from the roles list.",
                    })
                seen_roles.add(role)
                declared_roles.add(role)

    # 7. Semantic Checks: Capabilities & Connectors
    capabilities = manifest.get("capabilities")
    if isinstance(capabilities, dict):
        connectors = capabilities.get("connectors")
        if isinstance(connectors, list):
            for i, conn in enumerate(connectors):
                if isinstance(conn, dict):
                    name = conn.get("name", f"connector_{i}")
                    identity = conn.get("identity", "viewer")

                    if identity not in ("viewer", "service"):
                        checks.append({
                            "name": f"connector_identity_{name}",
                            "status": "fail",
                            "code": "invalid_manifest",
                            "path": f"capabilities.connectors.{i}.identity",
                            "message": f"Invalid connector identity '{identity}'.",
                            "hint": "Connector identity must be either 'viewer' or 'service'.",
                        })
                    elif identity == "service":
                        required_approvals.append(f"connector:{name}:service_identity")
                        checks.append({
                            "name": f"connector_approval_{name}",
                            "status": "fail",
                            "code": "approval_required",
                            "path": f"capabilities.connectors.{i}.identity",
                            "message": f"Connector '{name}' requests 'service' identity which requires owner approval.",
                            "hint": "An organization administrator or app owner must approve 'service' identity.",
                        })

                    # Role check
                    req_roles = conn.get("required_roles")
                    if isinstance(req_roles, list):
                        for r in req_roles:
                            if r not in declared_roles:
                                checks.append({
                                    "name": f"connector_role_{name}_{r}",
                                    "status": "fail",
                                    "code": "invalid_manifest",
                                    "path": f"capabilities.connectors.{i}.required_roles",
                                    "message": f"Connector '{name}' requires undeclared role '{r}'.",
                                    "hint": f"Add '{r}' to the top-level 'roles' list in capsule.manifest.yaml.",
                                })

    # 8. Semantic Checks: Egress
    egress = manifest.get("egress")
    if egress is None or len(egress) == 0:
        checks.append({
            "name": "egress_default_deny",
            "status": "pass",
            "code": "egress_denied_default",
            "path": "egress",
            "message": "Outbound egress is denied by default.",
        })
    elif isinstance(egress, list):
        for i, entry in enumerate(egress):
            if isinstance(entry, dict):
                host = entry.get("host")
                if not host or not isinstance(host, str):
                    checks.append({
                        "name": f"egress_entry_{i}",
                        "status": "fail",
                        "code": "invalid_manifest",
                        "path": f"egress.{i}.host",
                        "message": "Egress entry missing host.",
                        "hint": "Provide a valid FQDN in the 'host' field (e.g. api.github.com).",
                    })
                    continue

                if "*" in host:
                    checks.append({
                        "name": f"egress_wildcard_{i}",
                        "status": "fail",
                        "code": "egress_denied",
                        "path": f"egress.{i}.host",
                        "message": f"Wildcard host '{host}' is prohibited.",
                        "hint": "Specify exact fully qualified domain names (e.g. api.github.com).",
                    })
                elif IP_REGEX.match(host):
                    checks.append({
                        "name": f"egress_ip_{i}",
                        "status": "fail",
                        "code": "egress_denied",
                        "path": f"egress.{i}.host",
                        "message": f"Direct IP address '{host}' is prohibited.",
                        "hint": "Use fully qualified domain names instead of IP addresses.",
                    })
                elif not HOSTNAME_REGEX.match(host):
                    checks.append({
                        "name": f"egress_format_{i}",
                        "status": "fail",
                        "code": "egress_denied",
                        "path": f"egress.{i}.host",
                        "message": f"Host '{host}' is not a valid FQDN.",
                        "hint": "Do not include protocols (https://) or paths. Use format: hostname.domain.tld",
                    })
                else:
                    checks.append({
                        "name": f"egress_entry_{i}",
                        "status": "pass",
                        "code": "egress_allowed",
                        "path": f"egress.{i}.host",
                        "message": f"Egress host '{host}' format valid.",
                    })
        warnings.append("Non-empty egress declaration requires administrative monitoring.")

    # 9. Semantic Checks: Resource Limits
    limits = manifest.get("limits")
    if isinstance(limits, dict):
        mem = limits.get("memory_mb")
        if mem is not None and isinstance(mem, (int, float)) and mem > 512:
            checks.append({
                "name": "limit_memory",
                "status": "fail",
                "code": "quota_exceeded",
                "path": "limits.memory_mb",
                "message": f"Requested memory {mem}MB exceeds Alpha quota of 512MB.",
                "hint": "Reduce 'limits.memory_mb' to 512 or less.",
            })

        timeout = limits.get("request_timeout_s")
        if timeout is not None and isinstance(timeout, (int, float)) and timeout > 60:
            checks.append({
                "name": "limit_timeout",
                "status": "fail",
                "code": "quota_exceeded",
                "path": "limits.request_timeout_s",
                "message": f"Requested timeout {timeout}s exceeds Alpha quota of 60s.",
                "hint": "Reduce 'limits.request_timeout_s' to 60 or less.",
            })

        instances = limits.get("max_active_instances")
        if instances is not None and isinstance(instances, int) and instances > 1:
            checks.append({
                "name": "limit_instances",
                "status": "fail",
                "code": "quota_exceeded",
                "path": "limits.max_active_instances",
                "message": f"Requested instances {instances} exceeds Alpha maximum of 1.",
                "hint": "Set 'limits.max_active_instances' to 1 for Alpha single-instance execution.",
            })

    # Overall validity: no failed checks
    errors = [c for c in checks if c["status"] == "fail"]
    valid = len(errors) == 0

    effective_manifest = None
    if valid:
        normalized_connectors = []
        if isinstance(manifest.get("capabilities"), dict) and isinstance(manifest["capabilities"].get("connectors"), list):
            for c in manifest["capabilities"]["connectors"]:
                nc = dict(c)
                nc.setdefault("identity", "viewer")
                normalized_connectors.append(nc)

        effective_manifest = {
            "apiVersion": manifest.get("apiVersion", "capsule/v1alpha1"),
            "id": manifest.get("id"),
            "name": manifest.get("name"),
            "shape": manifest.get("shape", "web-app"),
            "runtime": manifest.get("runtime", "node22"),
            "roles": list(manifest.get("roles", [])),
            "capabilities": {
                **manifest.get("capabilities", {}),
                "connectors": normalized_connectors,
            } if manifest.get("capabilities") else None,
            "egress": list(manifest.get("egress", [])),
            "sharing": manifest.get("sharing", {"default": "org"}),
            "limits": {
                "cpu": manifest.get("limits", {}).get("cpu", "small"),
                "memory_mb": manifest.get("limits", {}).get("memory_mb", 256),
                "request_timeout_s": manifest.get("limits", {}).get("request_timeout_s", 30),
                "db_max_mb": manifest.get("limits", {}).get("db_max_mb", 50),
                "blob_max_mb": manifest.get("limits", {}).get("blob_max_mb", 200),
                "max_active_instances": 1,
            },

        }

    return {
        "valid": valid,
        "errors": errors,
        "checks": checks,
        "required_approvals": required_approvals,
        "warnings": warnings,
        "effective_manifest": effective_manifest,
    }
